Generic IManagerProvider and IManagerMetadata

jpinz opened the pull request:
Implement an `IManagerProvider` and an `IManagerMetadata` interface that gets created by an `IManagerProviderFactory` implemented by `ProviderFactory`.
Eventually a lot of the logic for handling manager specific functionality can be split into it's own projects/libraries/repos/etc like the NuGet.Protocol does for NuGet.

gfs commented on src/Shared.CLI/Helpers/PackageDownloader.cs:
```diff
@@ -23,9 +24,10 @@ public class PackageDownloader
         /// <param name="destinationDir">The directory where the package needs to be downloaded to.</param>
         /// <param name="doCaching">Check and use the cache if it exists - create if not.</param>
-        public PackageDownloader(PackageURL purl, IHttpClientFactory? httpClientFactory, string? destinationDir = null, bool doCaching = false)
+        public PackageDownloader(PackageURL purl, IHttpClientFactory? httpClientFactory, IManagerProvider<IManagerMetadata>? provider = null, string? destinationDir = null, bool doCaching = false)
```
Do we really need both an IHttpClientFactory and a IManagerProvider? Should the requests not flow through the ManagerProvider (which might itself take an IHttpClientFactory depending on the implementation).

jpinz replied:
That's a good point. I hadn't considered putting the IHttpClientFactory in the provider. I will give that a shot.

gfs commented on src/Shared/PackageManagers/PackageManagerFactory.cs:
```diff
@@ -16,29 +18,30 @@ public static class ProjectManagerFactory
         /// <param name="destinationDirectory">The directory to use to store any downloaded packages.</param>
         /// <returns> BaseProjectManager object </returns>
-        public static BaseProjectManager? CreateProjectManager(PackageURL purl, IHttpClientFactory? httpClientFactory = null, string? destinationDirectory = null)
+        public static BaseProjectManager? CreateProjectManager(PackageURL purl, IHttpClientFactory? httpClientFactory = null, IManagerProvider<IManagerMetadata>? managerProvider = null, string? destinationDirectory = null)
```
I don't understand how this pattern where you provide the provider here is going to work. It means you have to know ahead of time which manager your purl associates with, which doesn't fit this API surface - which has always been designed to take an arbitrary purl.

jpinz replied:
So my reasoning behind this, similarly to providing the IHttpClientFactory, is that most of the time a provider will not be provided. Just like most of the time we don't provide an IHttpClientFactory, except for our consumption case where we want to control the factory.
By providing the provider in the constructor, it allows us to provide a mocked implementation of the interface for unit test calls. There could very well be a better way of implementing this, but from my still somewhat limited understanding, we need to provide it here. 
I suppose one alternative is that I could add a method to the ProjectManagers that allows us to override and replace the provider that gets added during the factory construction (not from the parameters, but from the ProviderFactory) and then only use that method in unit tests?
If you know of any good design patterns for what I'm trying to accomplish please let me know. It really is only there for the unit test mocking, there shouldn't be a reason from any consuming level side for the consumer to want to override the provider like they might want to with the IHttpClientFactory.

gfs replied:
What about this:

Add an `IManagerProviderFactory` which takes a purl and an `IHttpClientFactory` and produces the correct `IManagerProvider`.
Each `BaseProjectManager` should have an associated `IManagerProvider`.
Instead of passing an `IHttpClientFactory` pass the `IManagerProviderFactory`.
Optional: the `IManagerProvider` interface should expose the `IHttpClientFactory` that was provided to its factory when it was constructed as well.
When passed a null `IManagerProviderFactory` instantiate the default ones -  (keep same default behavior).
When running unit tests Mock the `IManagerProviderFactory` to produce `IManagerProvider` mocks.

gfs replied:
This keeps the same flexibility as well to override the `IHttpClientFactory` as before without worrying about writing the `IManagerProvider` as the factory can pass in your `IHttpClientFactory` if you desire.

gfs replied:
It looks like you may have started a bit with that on the `ProviderFactory.cs`. I think you should just fully commit to the `IProvider` pattern and replace the `HttpClientFactory` parameters entirely (on the managers), with that instead being (optionally) passed to the factory.

jpinz replied:
Just made a huge refactor that did all of that. I think it accomplishes what you were thinking of.

gfs reviewed (changes requested):
I see a couple issues here.

1. I'm not sure why you would need to provide both an IHttpClient and an IManagerProvider. If the goal is to replace the amnager providers with individual implementations it should be the only parameter and the IHttpClientFactory should pass into the manager when needed.
2. I don't think these new proposed public APIs make sense. This is most obvious with the packagemanagerfactory but you can also see it with the tools. The CLI tools and createprojectmanager method are designed to take any arbitrary purl, but if you are providing them the ManagerProvider it will obviously only work with a matching purl.  I think this needs a rethink on how you get the manager provider.  Maybe it should be a manager provider factory that can produce the manager provider with your selected settings based on the purl?

gfs commented:
Will review refactor tomorrow.

brbayes-msft commented on src/Shared/Model/Providers/ProviderFactory.cs:
```diff
@@ -0,0 +1,68 @@
+    {
+        if (ManagerProviders.Count == 0)
+        {
+            ManagerProviders.AddRange(typeof(BaseProvider).Assembly.GetTypes()
```
Could use another reviewer here to double check that this isn't at risk for malicious implementations being injected at runtime.  I think it's okay because it's limited to the same assembly as BaseProvider, but could use someone else to double check.

gfs replied:
I am not a fan of this implicit enumeration - I think it would make sense to have the list of available providers set before you call this method - it seems more appropriate for the constructor - perhaps also as an optional parameter to skip this reflection.

brbayes-msft commented on src/Shared/PackageManagers/PackageManagerFactory.cs:
```diff
@@ -28,50 +30,39 @@ public static BaseProjectManager CreateBaseProjectManager(IHttpClientFactory htt
         public static BaseProjectManager CreateBaseProjectManager(string destinationDirectory)
         {
-            return new BaseProjectManager(destinationDirectory);
+            return new BaseProjectManager(new BaseProvider(), destinationDirectory);
```
If the intent here is to have a provider that complies with the interface but does not actually return any data, perhaps either a) the BaseProjectManager could be modified to allow the provider to be nullable, returning NotSupported if any relevant calls are made, or b) a NoOpProvider could be added instead of relying on functionality in a base class that does the same thing as the base class right now, but can instead be sealed so that no other classes try to extend it.

gfs replied:
The only reference in the lib I can find to this is from DetectCryptographyTool. I too am not quite sure why you would want this interface instead of the one based on purl.

jpinz replied:
Removed

brbayes-msft commented on src/Shared/PackageManagers/NuGetProjectManager.cs:
```diff
@@ -84,16 +83,19 @@ private async Task<string> GetRegistrationEndpointAsync()
 
-            string? packageName = purl?.Name;
-            string? packageVersion = purl?.Version;
+            string? packageName = purl.Name;
```
Is the PackageUrl no longer nullable here?

jpinz replied:
It isn't. You need a PackageUrl with a name and version specified in order to download said package version.

gfs replied:
I think the point was that the purl itself is coming from a public api - even though you have declared it to not be null in the interface you can't trust that a user won't give you a null.

To be fair though it would actually crash on line 95 when you print the name wihout the null guard.

brbayes-msft commented on src/Shared/PackageManagers/NuGetProjectManager.cs:
```diff
@@ -208,77 +140,26 @@ public override async Task<bool> PackageExists(PackageURL purl, bool useCache =
+            Logger.Trace("EnumerateVersions {0}", purl.ToString());
 
-            if (purl == null || purl.Name is null)
+            if (purl.Name is null)
```
Shouldn't this be `purl?.Name is null`?

jpinz replied:
purl shouldn't ever be null. I did just change it though. Now it is `if (string.IsNullOrWhiteSpace(purl.Name))`

brbayes-msft commented on src/Shared/PackageManagers/NuGetProjectManager.cs:
```diff
@@ -337,267 +210,199 @@ public override Uri GetPackageAbsoluteUri(PackageURL purl)
-            catch (InvalidOperationException) { return null; }
 
-            return null;
+            // TODO: Maintainers
```
Please write out this todo as a sentence.

gfs commented on src/Shared/Contracts/IManagerProvider.cs:
```diff
@@ -0,0 +1,82 @@
+    /// <summary>
+    /// The <see cref="IHttpClientFactory"/> for this <see cref="IManagerProvider{T}"/> to make <see cref="HttpClient"/>s with.
+    /// </summary>
+    public IHttpClientFactory HttpClientFactory { get; }
```
Since many of the managers may use HttpClientFactory behind the scenes I think it makes sense to have an `IHttpClientBasedManagerProvider` or `BaseHttpClientBasedManagerProvider` that does have this field to make it easier to develop those. But agree the base interface does not need it.

jpinz replied:
Removed HttpClientFactory from the IManagerProvider.

gfs commented on src/Shared/Contracts/IManagerProvider.cs:
```diff
@@ -0,0 +1,82 @@
+/// The interface to implement project manager specific methods.
+/// Useful for mocking when doing unit tests.
+/// </summary>
+public interface IManagerProvider<T>
```
Agree, I don't think this needs to be generic.

gfs commented on src/Shared/Model/Metadata/NuGetProtocolMetadata.cs:
```diff
@@ -0,0 +1,166 @@
+    /// Serialize this instance of <see cref="NuGetMetadata"/> into a json string.
+    /// </summary>
+    /// <returns>A json string representing this object.</returns>
+    public string ToJson()
```
Agree, I don't think this is needed. There are no special serialization options being used.

gfs commented on src/Shared/Model/Providers/NuGetProvider.cs:
```diff
@@ -0,0 +1,150 @@
+
+    /// <inheritdoc />
+    public override async Task<string?> DownloadAsync(BaseProjectManager projectManager, PackageURL packageUrl, string targetDirectory, bool doExtract,
+        bool cached = false, CancellationToken cancellationToken = default)
```
I don't think this should take a project manager as an argument.  It doesn't appear to actually need the manager.

The only think it used for is `ExtractArchive` which is actually a method implemented on the base class as far as I can tell which just leverages `RecursiveExtractor` and could be refactored into a static helper method.

jpinz replied:
Made it a static method now. Still in BaseProjectManager though. Maybe I should make an ArchiveHelper class? Thoughts?

gfs commented on src/Shared/Model/Providers/ProviderFactory.cs:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    // do reflection only once
+    private static readonly List<Type> ManagerProviders = new();
```
This reflection is actually done once per provider factory. It may make sense to do this reflection one time and pass the providers you want in instead when constructing the provider factory.

gfs commented on src/Shared/Model/Providers/ProviderFactory.cs:
```diff
@@ -0,0 +1,68 @@
+        if (ctor != null)
+        {
+            BaseProvider? provider = (BaseProvider)ctor.Invoke(new object?[] { HttpClientFactory });
+            return provider;
```
Since we are caching the results of reflection shouldn't we cache this as well?

gfs commented on src/Shared/PackageManagers/PackageManagerFactory.cs:
```diff
@@ -28,50 +30,39 @@ public static BaseProjectManager CreateBaseProjectManager(IHttpClientFactory htt
-               .Where(type => type.Name.Equals($"{purl.Type}ProjectManager",
-                                               StringComparison.InvariantCultureIgnoreCase))
-               .FirstOrDefault();
+            Type? managerClass = projectManagers
```
This reflection is potentially run many times.

gfs reviewed (changes requested):
See comments throughout and agree with the other comments left as well.

jpinz commented on src/oss-find-squats-lib/MutateExtension.cs:
```diff
@@ -113,8 +114,8 @@ public static class MutateExtension
                     // Construct the mutated name if the package was scoped.
-                    string? mutationNamespace = isScoped ? mutation.Mutated : null;
-                    string mutationName = isScoped ? purl.Name : mutation.Mutated;
+                    string? mutationNamespace = HttpUtility.UrlEncode(isScoped ? mutation.Mutated : null);
```
I had a case where `newtonsoft.json` was mutated into `newtonsoft/json`, which when parsed back into a package url became `pkg:nuget/newtonsoft/json` which would be interpreted as the namespace being `newtonsoft`, and the name being `json`. We might want to change that on the mutator level to be impossible to do, but this also fixes it.

jpinz commented on src/Shared/Model/Providers/BaseProvider.cs:
```diff
@@ -0,0 +1,50 @@
+    /// <inheritdoc />
+    public virtual Task<string?> DownloadAsync(BaseProjectManager projectManager, PackageURL packageUrl, string targetDirectory, bool doExtract,
+        bool cached = false, CancellationToken cancellationToken = default) =>
+        throw new System.NotImplementedException();
```
Good call!

jpinz commented on src/Shared/Model/Providers/NuGetProvider.cs:
```diff
@@ -0,0 +1,150 @@
+    /// <summary>
+    /// Instantiates a new <see cref="NuGetProvider"/>.
+    /// </summary>
+    public NuGetProvider(IHttpClientFactory? httpClientFactory = null)
```
done!

gfs commented on src/Shared/PackageManagers/NuGetProjectManager.cs:
```diff
@@ -84,16 +83,19 @@ private async Task<string> GetRegistrationEndpointAsync()
         public override async Task<IEnumerable<string>> DownloadVersion(PackageURL purl, bool doExtract, bool cached = false)
         {
-            Logger.Trace("DownloadVersion {0}", purl?.ToString());
+            Logger.Trace("DownloadVersion {0}", purl.ToString());
```
In C#11 you will be able to do this instead with:
```
public override async Task<IEnumerable<string>> DownloadVersion(PackageURL purl!!, bool doExtract, bool cached = false)
```

gfs replied:
Actually, we could enable `!!` by adding `<EnablePreviewFeatures>True</EnablePreviewFeatures>` to the csproj.

https://dotnetcoretutorials.com/2022/02/18/c11-parameter-null-checking/

jpinz replied:
I like that, will probably do that in a separate PR.

jpinz commented on src/Shared/Model/Providers/BaseProvider.cs:
```diff
@@ -0,0 +1,50 @@
+using System.Threading;
+using System.Threading.Tasks;
+
+public class BaseProvider : IManagerProvider<IManagerMetadata>
```
Class was removed

jpinz commented on src/Shared/Model/Providers/BaseProvider.cs:
```diff
@@ -0,0 +1,50 @@
+using System.Threading;
+using System.Threading.Tasks;
+
+public class BaseProvider : IManagerProvider<IManagerMetadata>
```
Class removed. NoOpProvider covers what I was using this for.

jpinz commented on src/Shared/PackageManagers/BaseProjectManager.cs:
```diff
@@ -552,7 +556,7 @@ public virtual List<Version> GetVersions(JsonDocument? metadata)
         protected HttpClient CreateHttpClient()
         {
-            return HttpClientFactory.CreateClient(GetType().Name);
+            return Provider.CreateHttpClient(GetType().Name);
```
Yeah.

jpinz replied:
Removed HttpClientFactory from providers.

jpinz commented on src/Shared/Model/Providers/ProviderFactory.cs:
```diff
@@ -0,0 +1,68 @@
+            .FirstOrDefault(type => type.Name.Equals($"{purl.Type}Provider",
+                StringComparison.InvariantCultureIgnoreCase));
+
+        if (providerClass == null)
```
We now throw a new NotSupportedException(purl.Type) in this case.

jpinz commented on src/Shared/Model/Providers/NuGetProvider.cs:
```diff
@@ -0,0 +1,150 @@
+    /// <summary>
+    /// Instantiates a new <see cref="NuGetProvider"/>.
+    /// </summary>
+    public NuGetProvider(IHttpClientFactory? httpClientFactory = null)
```
NuGetProvider doesn't take in an IHttpClientFactory anymore.

gfs commented:
Let me know when you'd like me to review again.

jpinz commented on src/Shared/PackageManagers/PackageManagerFactory.cs:
```diff
@@ -28,50 +30,39 @@ public static BaseProjectManager CreateBaseProjectManager(IHttpClientFactory htt
-        /// <returns> BaseProjectManager object </returns>
-        public static BaseProjectManager? CreateProjectManager(PackageURL purl, IHttpClientFactory? httpClientFactory = null, string? destinationDirectory = null)
+        /// <returns>A new <see cref="BaseProjectManager"/> object implementation.</returns>
+        public static BaseProjectManager? CreateProjectManager(PackageURL purl, IManagerProviderFactory? managerProviderFactory = null, string? destinationDirectory = null)
```
Renamed to ProjectManagerFactory.
The IManagerProviderFactory is now gone, I refactored it all again.

gfs commented:
Changes are major enough that I will need to pull down and poke around in my IDE.  Will review tomorrow.

pmalmsten commented on src/Shared/PackageManagers/ProjectManagerFactory.cs:
```diff
@@ -0,0 +1,139 @@
+        private readonly IHttpClientFactory _httpClientFactory;
+        private readonly IManagerProvider? _nuGetProvider;
+        
+        /// <summary>
+        /// The dictionary of project managers.
+        /// </summary>
+        private static readonly Dictionary<string, ConstructProjectManager> ProjectManagers =
+            new(StringComparer.InvariantCultureIgnoreCase)
```
As we discussed offline, if this field were non-static and initialized in the constructor instead, the lambdas here would be able to directly access the singleton services they need instead of needing to pass them in via the `ConstructProjectManager` delegate.

By doing that, we could drop the two private fields (for `IHttpClientFactory` and `IManagerProvider?`, drop the corresponding two arguments from `ConstructProjectManager`, etc.

pmalmsten commented on src/Shared/Contracts/IManagerProvider.cs:
```diff
@@ -0,0 +1,66 @@
+/// The interface to implement project manager specific methods.
+/// Useful for mocking when doing unit tests.
+/// </summary>
+public interface IManagerProvider
```
Nit: the name `IManagerProvider` sounds to me like something that can provide Managers (e.g. a manager factory). That doesn't really seem like what implementations of this type is doing - it seems like the methods here are more focused on particular package actions? 

I wonder if there could be a better name that conveys that more clearly - as a brainstorm, would something like `IManagerPackageActions` make sense?

jpinz replied:
@brbayes-msft and I were chatting about this. We didn't have a better name at the time. I like your suggestion.
Any thoughts @gfs ?

jpinz replied:
I renamed it to `IManagerPackageActions`, we can change it pretty easily if we decide otherwise. But it makes a lot more sense than `IManagerProvider`

pmalmsten commented on src/Shared/Model/Metadata/NuGetProtocolMetadata.cs:
```diff
@@ -0,0 +1,145 @@
+/// <summary>
+/// A class to represent Package Metadata for a NuGet package.
+/// </summary>
+public record NuGetMetadata : IManagerMetadata
```
Nit: should we emphasize the package-version-specific nature of this metadata? Maybe call it something like `NuGetPackageVersionMetadata`?

jpinz replied:
I renamed it, and I ended up renaming the interface to `IManagerPackageVersionMetadata`

pmalmsten commented on src/Shared/Model/Providers/NuGetProvider.cs:
```diff
@@ -0,0 +1,162 @@
+    private readonly SourceRepository _sourceRepository = Repository.Factory.GetCoreV3("https://api.nuget.org/v3/index.json");
+    private readonly NuGetLogger _logger = new(LogManager.GetCurrentClassLogger());
+
+    /// <summary>
+    /// Instantiates a new <see cref="NuGetProvider"/>.
+    /// </summary>
+    public NuGetProvider()
+    {
+    }
```
Nit: this looks equivalent to the default constructor that C# automatically provides. Should we delete this in favor of the default constructor?

pmalmsten commented on src/Shared/Model/Providers/NuGetProvider.cs:
```diff
@@ -0,0 +1,162 @@
+    {
+        FindPackageByIdResource resource = await _sourceRepository.GetResourceAsync<FindPackageByIdResource>();
+
+        // Create a new memory stream to populate with the .nupkg.
+        using MemoryStream packageStream = new();
```
How large can NuGet package files get? Have we considered something like downloading to a temporary file instead of writing to RAM?

pmalmsten replied:
Looks like there are some decent ways to go about using a temporary file: https://stackoverflow.com/questions/34776031/how-to-create-a-temp-file-that-is-automatically-deleted-after-the-program-termin

pmalmsten replied:
This point could easily be addressed in a separate PR though - not a blocker here.

jpinz replied:
Probably will address in another PR. But this implementation is straight from the documentation just as an FYI:
https://docs.microsoft.com/en-us/nuget/reference/nuget-client-sdk#download-a-package

jpinz replied:
Addressed.

pmalmsten commented on src/Shared/PackageManagers/NuGetProjectManager.cs:
```diff
@@ -2,39 +2,42 @@
+        private IManagerProvider ManagerProvider { get; }
 
-        public NuGetProjectManager(IHttpClientFactory httpClientFactory, string destinationDirectory) : base(httpClientFactory, destinationDirectory)
+        public NuGetProjectManager(IHttpClientFactory httpClientFactory, string destinationDirectory, IManagerProvider? provider = null) : base(httpClientFactory, destinationDirectory)
         {
+            ManagerProvider = provider ?? new NuGetProvider();
             GetRegistrationEndpointAsync().Wait();
         }
 
         public NuGetProjectManager(string destinationDirectory) : base(destinationDirectory)
         {
+            ManagerProvider = new NuGetProvider();
             GetRegistrationEndpointAsync().Wait();
         }
```
Nit: Could we simply make the `IManagerProvider` required? It seems like the only existing uses always supply the argument.

pmalmsten commented on src/Shared/PackageManagers/NuGetProjectManager.cs:
```diff
@@ -2,39 +2,42 @@
 
         private string? RegistrationEndpoint { get; set; } = null;
+        private IManagerProvider ManagerProvider { get; }
 
-        public NuGetProjectManager(IHttpClientFactory httpClientFactory, string destinationDirectory) : base(httpClientFactory, destinationDirectory)
+        public NuGetProjectManager(IHttpClientFactory httpClientFactory, string destinationDirectory, IManagerProvider? provider = null) : base(httpClientFactory, destinationDirectory)
```
Will this class function properly if any other implementation of `IManagerProvider` is provided? If not, should we use a narrower interface type that only `NuGetProvider` implements? 

(The fundamental idea of interfaces is to allow for interchanging of different implementations that all satisfy the same contract - if this class would not work if (say) a future `NpmProvider` was provided here, then we should have a narrower interface type specific just to NuGet that this class accepts instead. Such a narrower interface type could extend from `IManagerProvider` if we want to share common method signatures).

gfs replied:
It should be `IManagerProvider<NugetProviderMetadata>` I think.

pmalmsten commented on src/Shared/PackageManagers/NuGetProjectManager.cs:
```diff
@@ -292,39 +178,31 @@ public override async Task<IEnumerable<string>> EnumerateVersions(PackageURL pur
+                // Construct a new PackageURL that's guaranteed to have a  version.
+                PackageURL purlWithVersion = new (purl.Type, purl.Namespace, packageName, packageVersion, purl.Qualifiers, purl.Subpath);
+                
+                NuGetMetadata? packageVersionMetadata =
+                    await ManagerProvider.GetMetadataAsync(purlWithVersion, useCache: useCache) as NuGetMetadata;
```
Regarding casting the result to `NuGetMetadata` - this goes with my earlier comment about other implementations of `IManagerProvider` not necessarily being suitable for use with this class. If we had a Nuget-specific version of `IManagerProvider`, that would eliminate the need for casting the return value here, since the Nuget-specific `IManagerProvider` would just return `NuGetMetadata` directly.

jpinz commented on src/Shared/Contracts/IManagerMetadata.cs:
```diff
@@ -0,0 +1,16 @@
+
+namespace Microsoft.CST.OpenSource.Contracts;
+
+public interface IManagerMetadata
```
I could add some documentation for it. But the idea here is this is the interface that ProjectManagers will implement for whatever metadata is returned by the registry API. It won't contain all of the information we want, just whatever the registry returns. So for example NuGet's current implementation doesn't contain a URI to download the nupkg, or any digest for the package, but we want that data. It just doesn't get returned from the API we are using.
So I think the interface name is appropriate, but I will add some docs to explain it better.

jpinz replied:
Here's what I have right now:
"An interface representing the metadata returned from a project manager's API for metadata, won't contain extra or calculated properties."

jpinz replied:
I changed my mind and renamed it to `IManagerPackageVersionMetadata` it's a bit wordy, but makes more sense.

pmalmsten replied:
`IManagerPackageVersionMetadata` sounds pretty decent to me.

pmalmsten commented on src/Shared/Model/Metadata/NuGetPackageVersionPackageVersionMetadata.cs:
```diff
@@ -13,9 +13,9 @@ namespace Microsoft.CST.OpenSource.Model.Metadata;
+/// A class to represent Package Metadata for a NuGet package version.
 /// </summary>
-public record NuGetMetadata : IManagerMetadata
+public record NuGetPackageVersionPackageVersionMetadata : IManagerPackageVersionMetadata
```
Looks like this was fixed in the most recent update actually - resolving this thread.

pmalmsten commented on src/Shared/PackageManagers/NuGetProjectManager.cs:
```diff
@@ -292,39 +178,31 @@ public override async Task<IEnumerable<string>> EnumerateVersions(PackageURL pur
+                PackageURL purlWithVersion = new (purl.Type, purl.Namespace, packageName, packageVersion, purl.Qualifiers, purl.Subpath);
+                
+                NuGetPackageVersionMetadata? packageVersionMetadata =
+                    await NuGetPackageActions.GetMetadataAsync(purlWithVersion, useCache: useCache) as NuGetPackageVersionMetadata;
```
Nit: casting is no longer required now that the interface is generic.

Same note applies in a few places in this file.

pmalmsten reviewed (commented):
Recent changes look pretty good - just gave you one more nit.

gfs commented on src/Shared/PackageManagers/CPANProjectManager.cs:
```diff
@@ -120,7 +120,7 @@ public override async Task<IEnumerable<string>> DownloadVersion(PackageURL purl,
                 if (doExtract)
                 {
-                    downloadedPaths.Add(await ExtractArchive(targetName, await result.Content.ReadAsByteArrayAsync(), cached));
+                    downloadedPaths.Add(await ExtractArchive(TopLevelExtractionDirectory, targetName, await result.Content.ReadAsByteArrayAsync(), cached));
```
Why are we converting the Stream to an array? I think this work if you just pass the stream instead of reading hteb ytes and then making a new stream to hold it.

gfs replied:
If that has not been working with RecursiveExtractor let me know and I can investigate.

jpinz replied:
I think it works fine now.

gfs commented on src/Shared/PackageManagers/BaseProjectManager.cs:
```diff
@@ -323,7 +324,7 @@ public async Task<string> ExtractArchive(string directoryName, byte[] bytes, boo
                 // MaxExtractedBytes = 1000 * 1000 * 10;  // 10 MB maximum package size
             };
-            ExtractionStatusCode result = await extractor.ExtractToDirectoryAsync(TopLevelExtractionDirectory, dirBuilder.ToString(), new MemoryStream(bytes), extractorOptions);
+            ExtractionStatusCode result = await extractor.ExtractToDirectoryAsync(topLevelDirectory, dirBuilder.ToString(), new MemoryStream(bytes), extractorOptions);
```
You end up turning the bytes back into a Stream before calling extract anyway. Recommend you just pass the Stream.

jpinz replied:
Switched to passing the stream around instead.

gfs commented on src/Shared/PackageManagers/NuGetProjectManager.cs:
```diff
@@ -2,39 +2,42 @@
-        public NuGetProjectManager(IHttpClientFactory httpClientFactory, string destinationDirectory) : base(httpClientFactory, destinationDirectory)
+        public NuGetProjectManager(IHttpClientFactory httpClientFactory, string destinationDirectory, NuGetPackageActions nugetPackageActions) : base(httpClientFactory, destinationDirectory)
         {
+            NuGetPackageActions = nugetPackageActions;
```
Class name and Instance variable of that class are named identically.  This is a private (also probably readonly?) so naming pattern would normally be `_nuGetPackageActions`

gfs commented on src/Shared/PackageManagers/NuGetProjectManager.cs:
```diff
@@ -337,267 +215,199 @@ public override Uri GetPackageAbsoluteUri(PackageURL purl)
-            // Title is different than name or description for NuGet packages, not always used.
-            // metadata.Title = GetLatestCatalogEntry(root).GetProperty("title").GetString();
+            NuGetPackageVersionMetadata packageVersionPackageVersionMetadata =
+                await NuGetPackageActions.GetMetadataAsync(purlWithVersion, useCache: useCache) as NuGetPackageVersionMetadata ?? throw new InvalidOperationException($"The metadata for {purlWithVersion} was null.");
```
Recommend returning `null` instead of throwing here.

gfs commented on src/Shared/PackageManagers/ProjectManagerFactory.cs:
```diff
@@ -0,0 +1,128 @@
+        /// </summary>
+        /// <param name="httpClientFactoryParam">The <see cref="IHttpClientFactory"/> to use in the project managers.</param>
+        /// <param name="nugetPackageActions">The <see cref="NuGetPackageActions"/> to use in the <see cref="NuGetProjectManager"/>.</param>
+        public ProjectManagerFactory(IHttpClientFactory? httpClientFactoryParam = null, NuGetPackageActions? nugetPackageActions = null)
```
```suggestion
        public ProjectManagerFactory(IHttpClientFactory? httpClientFactoryParam = null, IManagerPackageActions? managerPackageActions = null)
```

jpinz replied:
The issue with the factory now is that it's not exactly DI, but trying to mimic how DI would work. So in that case, we would need one parameter for each project manager that uses a `IManagerPackageActions` in the ProjectManager implementation. So say one gets added for NPM, the `ProjectManagerFactory` constructor would have to be:
`public ProjectManagerFactory(IHttpClientFactory? httpClientFactoryParam = null, NuGetPackageActions? nugetPackageActions = null, NpmPackageActions? npmPackageActions = null)` so that way we can override one, the other, both, or neither.

jpinz replied:
In other words, I don't think we can change it to just `managerPackageActions`, as it will have to be manager specific at least given this new design.

gfs replied:
Given that the ctor has these tightly coupled arguments I don't think this is a good choice to use for the public API for the reasons you mention. This simply does not scale.

Here's my proposal similar to before, to help avoid some of this reflection:

I think this class should have this ctor

```
public ProjectManagerFactory(Dictionary<string, ConstructProjectManager>? startingManagerSet = null)
{
   // Keep the existing default behavior if you don't provide any managers, otherwise use the set the user provided
    _projectManagers = startingManagerSet ?? GenerateDefaultSet();
...
}
```

```
public Dictionary<string, ConstructProjectManager> GenerateDefaultSet(IHttpClientFactory? factory = null)
{
// Existing code that generates the set below
}
```

It should have these methods that manipulate the dictionary
```
// Replace the manager for the specified purl type.
public bool SetManager(string purlType, ConstructProjectManager constructMethod)
// Replace the dictionary with the provided dictionary
public bool SetManagers(IDictionary<string, ConstructProjectManager> managersToReplace)
// Remove the manager for the specified purl type
public bool RemoveManager(string purlType)
// Remove all entries from the dictionary
public bool ClearManagers()
```

gfs commented on src/Shared/PackageManagers/ProjectManagerFactory.cs:
```diff
@@ -0,0 +1,128 @@
+                },
+                {
+                    nameof(NuGetProjectManager), destinationDirectory =>
+                        new NuGetProjectManager(httpClientFactory, destinationDirectory, nugetPackageActions ?? new NuGetPackageActions()) // Add the NuGetPackageActions to the NuGetProjectManager.
```
```suggestion
                        new NuGetProjectManager(httpClientFactory, destinationDirectory, managerPackageActions is null ? new NugetPackageActions() : managerPackageActions is NugetPackageActions npa ? npa : throw new ArgumentException(nameof(managerPackageActions)) // Add the NuGetPackageActions to the NuGetProjectManager.
```

jpinz replied:
I don't think this is applicable based on the comment I left: https://github.com/microsoft/OSSGadget/pull/318#discussion_r841916574

gfs commented on src/Shared/PackageManagers/ProjectManagerFactory.cs:
```diff
@@ -0,0 +1,128 @@
+        /// <param name="httpClientFactory">The <see cref="IHttpClientFactory"/> to optionally add.</param>
+        /// <param name="nugetPackageActions">The <see cref="NuGetPackageActions"/> to use in the <see cref="NuGetProjectManager"/>.</param>
+        /// <returns>A new <see cref="BaseProjectManager"/> implementation.</returns>
+        public static BaseProjectManager? GetProjectManager(PackageURL packageUrl, IHttpClientFactory? httpClientFactory = null, NuGetPackageActions? nugetPackageActions = null)
```
```suggestion
        public static BaseProjectManager? GetProjectManager(PackageURL packageUrl, IHttpClientFactory? httpClientFactory = null, IManagerPackageActions? managerPackageActions = null)
```

jpinz replied:
Also don't think this is applicable base on my previous comment: https://github.com/microsoft/OSSGadget/pull/318#discussion_r841916574

gfs replied:
See newer comment. This is not a scalable approach and must be changed.

gfs commented on src/oss-characteristics/CharacteristicTool.cs:
```diff
@@ -73,11 +74,15 @@ public static IEnumerable<Example> Examples
+        {
+        }
+
         public CharacteristicTool(IHttpClientFactory httpClientFactory) : base(httpClientFactory)
```
Remove

gfs replied:
I think we should be removing all the IHttpClient based ctors.

gfs replied:
Or like below it should make a projectmanagerfactory with that httpclient to maintain compat.

gfs commented on src/oss-characteristics/CharacteristicTool.cs:
```diff
@@ -73,11 +74,15 @@ public static IEnumerable<Example> Examples
+        {
+        }
+
         public CharacteristicTool(IHttpClientFactory httpClientFactory) : base(httpClientFactory)
```
```suggestion
        public CharacteristicTool(IHttpClientFactory httpClientFactory) : base(new ProjectManagerFactory(httpClientFactory))
```

gfs commented on src/oss-find-source/FindSourceTool.cs:
```diff
@@ -23,11 +24,15 @@ public class FindSourceTool : OSSGadget
+        {
+        }
+
         public FindSourceTool(IHttpClientFactory httpClientFactory) : base(httpClientFactory)
```
```suggestion
        public FindSourceTool(IHttpClientFactory httpClientFactory) : this(new ProjectManagerFactory(httpClientFactory))
```

gfs commented on src/oss-find-squats-lib/MutateExtension.cs:
```diff
@@ -11,6 +11,7 @@ namespace Microsoft.CST.OpenSource.FindSquats.ExtensionMethods
     using System.Collections.Generic;
     using System.Linq;
     using System.Threading;
+    using System.Web;
```
? Dependency change with no code change?

gfs commented on src/oss-health/HealthTool.cs:
```diff
@@ -17,22 +17,28 @@ namespace Microsoft.CST.OpenSource
+        {
+        }
+
         public HealthTool(IHttpClientFactory httpClientFactory) : base(httpClientFactory)
```
```suggestion
        public HealthTool(IHttpClientFactory httpClientFactory) : this(new ProjectManagerFactory(httpClientFactory))
```

gfs replied:
Unless there is something I'm missing this seems like a better pattern.  At the very least we need to choose one to use because there are both ones that case base with httpclientfactory and those that make a projectmanagerfactory based on the httpclient.

jpinz replied:
Yeah I switched all of the OSSGadget CLI tools to this pattern.

gfs commented on src/oss-reproducible/ReproducibleTool.cs:
```diff
@@ -75,12 +75,18 @@ public static IEnumerable<Example> Examples
+        {
+        }
+
         public ReproducibleTool(IHttpClientFactory httpClientFactory) : base(httpClientFactory)
```
+

gfs commented on src/oss-risk-calculator/RiskCalculatorTool.cs:
```diff
@@ -67,11 +68,15 @@ public static IEnumerable<Example> Examples
+        {
+        }
+
         public RiskCalculatorTool(IHttpClientFactory httpClientFactory) : base(httpClientFactory)
```
+

gfs commented on src/oss-tests/FindSquatsTests.cs:
```diff
@@ -23,7 +27,7 @@ namespace Microsoft.CST.OpenSource.Tests
     {
         [DataTestMethod]
-        [DataRow("pkg:nuget/Microsoft.CST.OAT", false)]
+        // [DataRow("pkg:nuget/Microsoft.CST.OAT", false)] TODO: Figure out why this fails on ADO.
```
Does this only fail in pipeline?

jpinz replied:
Yes. I have yet to be able to repro it locally.

jpinz commented on src/Shared/PackageManagers/ProjectManagerFactory.cs:
```diff
@@ -0,0 +1,139 @@
+                },
+            };
+
+        public ProjectManagerFactory(IHttpClientFactory? httpClientFactory = null, IManagerProvider? nuGetProvider = null)
```
ManagerProvider is no longer the naming scheme, it's now PackageActions. Did you review an old version of the code?

jpinz commented on src/Shared/PackageManagers/BaseProjectManager.cs:
```diff
@@ -289,15 +289,16 @@ public virtual Task<IEnumerable<string>> EnumerateVersions(PackageURL purl, bool
         /// <param name="cached">If the archive has been cached.</param>
         /// <returns></returns>
-        public async Task<string> ExtractArchive(string directoryName, byte[] bytes, bool cached = false)
+        public static async Task<string> ExtractArchive(string topLevelDirectory, string directoryName, byte[] bytes, bool cached = false)
```
Refactored to an ArchiveHelper.

jpinz commented on src/Shared.CLI/OSSGadget.cs:
```diff
@@ -21,6 +21,10 @@ public class OSSGadget : OssGadgetLib
+        {
+        }
+
         public OSSGadget(IHttpClientFactory httpClientFactory) : base(httpClientFactory)
```
Remove the ProjectManagerFactory one? or the IHttpClientFactory one?
I removed the IHttpClientFactory one based on some of your later comments.

gfs commented on src/Shared/PackageManagers/NuGetProjectManager.cs:
```diff
@@ -2,39 +2,42 @@
+        private NuGetPackageActions _nuGetPackageActions { get; }
 
-        public NuGetProjectManager(IHttpClientFactory httpClientFactory, string destinationDirectory) : base(httpClientFactory, destinationDirectory)
+        public NuGetProjectManager(IHttpClientFactory httpClientFactory, string destinationDirectory, NuGetPackageActions nugetPackageActions) : base(httpClientFactory, destinationDirectory)
```
```suggestion
        public NuGetProjectManager(IHttpClientFactory httpClientFactory, string destinationDirectory, IManagerPackageActions<NuGetPackageVersionMetadata> nugetPackageActions) : base(httpClientFactory, destinationDirectory)
```

gfs replied:
As long as something implements the manager actions that return nuget metadata we don't need to require it be the specific `NuGetPackageActions` implementation do we?

jpinz replied:
We don't, I changed it. Fixed the failing unit tests that way too! Thanks haha.

gfs commented on src/Shared/PackageManagers/NuGetProjectManager.cs:
```diff
@@ -208,77 +145,26 @@ public override async Task<bool> PackageExists(PackageURL purl, bool useCache =
-            if (purl == null || purl.Name is null)
+            if (string.IsNullOrWhiteSpace(purl.Name))
             {
                 return new List<string>();
```
```suggestion
                return Array.Empty<string>();
```
Less allocation overhead than providing a list.

gfs commented on src/oss-detect-cryptography/DetectCryptographyTool.cs:
```diff
@@ -92,9 +93,8 @@ private static async Task Main(string[] args)
-                            BaseProjectManager projectManager = ProjectManagerFactory.CreateBaseProjectManager(detectCryptographyTool.HttpClientFactory, targetDirectoryName);
 
-                            string? path = await projectManager.ExtractArchive("temp", File.ReadAllBytes(target));
+                            string? path = await ArchiveHelper.ExtractArchiveAsync(targetDirectoryName, "temp", File.OpenRead(target));
```
Nit: As part of the underlying implementation in RecursiveExtractor your provided stream will be used copied to a new Stream inside a FileEntry, because we can't make assumptions about the incoming stream - like that it can be seeked - but doing the extraction requires those properties.

I thought that the Filename based api avoid this extra allocation but it looks like it does not so I have a PR to improve it: https://github.com/microsoft/RecursiveExtractor/pull/77.

You can either call RecursiveExtractor's ExtractToDirectory with a filename or with FileEntry.FromFileName

jpinz replied:
> Some other misc issues.
> 
> 1. Does NugetPackageActions belong in the model directory? It seems like an implementation to me.
> 2. Why does NuGetPackageMetaData not derive from PackageMetadata?

The first one makes sense to me, I can refactor that.
NuGetPackageMetadata doesn't implement PackageMetadata as PackageMetadata was written as a "Catch-All" Metadata model, so it has unnecessary things that don't pertain to NuGet. The NuGetPackageMetadata in this case also doesn't represent what we want as an entire view of a NuGet package's metadata, currently what it is doing is just mapping the `PackageSearchMetadataRegistration` model from `NuGet.Protocol` to a model we control and can expand upon. Such as adding digest and hash data, plus some other data that we as a team have been discussing that we'd like as part of a package's metadata.
We're still flushing out final details, we'll loop you and @scovetta in once we're done to present to you both.

gfs commented on src/oss-diff/DiffTool.cs:
```diff
@@ -85,11 +85,15 @@ public static IEnumerable<Example> Examples
         }
 
-        public DiffTool() : this (new DefaultHttpClientFactory()) { }
+        public DiffTool(IHttpClientFactory httpClientFactory) : this(new ProjectManagerFactory(httpClientFactory))
```
I think all these ctors that take IHttpClientFactory should be removed.

jpinz commented on src/Shared/PackageManagers/BaseProjectManager.cs:
```diff
@@ -2,35 +2,33 @@
         /// Initializes a new instance of the <see cref="BaseProjectManager"/> class.
         /// </summary>
-        public BaseProjectManager(IHttpClientFactory httpClientFactory, string destinationDirectory)
+        public BaseProjectManager(IHttpClientFactory httpClientFactory, string destinationDirectory = ".")
```
No, because only NuGet right now uses any implementation of `IManagerPackageActions`, so there isn't a need to have it as part of the base project manager as well I don't think.
I separated the `IHttpClientFactory` out of the `IManagerPackageActions` (previously `IManagerProvider`) as it wasn't necessary to do so and caused some confusing scenarios where every project manager had a `IManagerPackageActions` but only used it as an object containing an `IHttpClientFactory`, adding an unnecessary wrapper essentially.

gfs replied:
> No, because only NuGet right now uses any implementation of `IManagerPackageActions`, so there isn't a need to have it as part of the base project manager as well I don't think.

If the purpose of the refactor was to have an interface for provider actions but you only use it for Nuget its not an interface.  Its just class with a complex type.

>  I separated the `IHttpClientFactory` out of the `IManagerPackageActions` (previously `IManagerProvider`) as it wasn't necessary to do so and caused some confusing scenarios where every project manager had a `IManagerPackageActions` but only used it as an object containing an `IHttpClientFactory`, adding an unnecessary wrapper essentially.

I think this is far more confusing.

jpinz replied:
I understand the confusion, I was struggling to decide how to handle it myself.
I personally think that the `IManagerPackageActions` shouldn't contain an `IHttpClientFactory`, at least not currently.
I think one thing that could be worth doing is a follow up PR that implements a new `IManagerPackageActions` class for each manager and then separate out the logic accordingly. I think then it would make sense to have the `IHttpClientFactory` as a part of `IManagerPackageActions` but we would also want it in the `ProjectManager` implementation as well I'd imagine.

gfs commented:
Some other misc issues.

1. Does NugetPackageActions belong in the model directory? It seems like an implementation to me.
2. Why does NuGetPackageMetaData not derive from PackageMetadata?

jpinz commented on src/Shared/Model/PackageActions/NuGetPackageActions.cs:
```diff
@@ -0,0 +1,156 @@
+        FindPackageByIdResource resource = await _sourceRepository.GetResourceAsync<FindPackageByIdResource>();
+
+        // Create a new memory stream to populate with the .nupkg.
+        using MemoryStream packageStream = new();
```
Good idea! Done.

gfs commented:
I did a partial implementation of the cargo manager in the proposed style that embraces the ManagerProvider interface.

https://github.com/microsoft/OSSGadget/pull/323/commits/16d089f66de6cfd1cc10f09da2f6b364a17bd833

In Doing so I noticed some potential concerns with that interface:
It returns `async Task<string>` what is the behavior supposed to be if you need to download multiple files?
Because it only returns string the doextract flag doesn't make sense.

Proposed either:

1. It should return IEnumerable<string> and keep the doextract flag
2. It keep returning string and remove the doextract flag
3. It returns bool and the location it is downloaded to is known by the caller because they passed it in.
4. The comments are changed to clarify that it returns a the path to a directory containing the all the downloaded files and not the files themselves.

jpinz commented:
> 

I don't believe I changed any logic from what is currently in the repo, but I can change it to reflect your points here.
Currently, it does 4, which I think is fine personally, but the documentation should definitely be better to reflect that.

gfs commented:
Got it. Lets leave it as 4 then.

jpinz commented on src/Shared/Model/Metadata/NuGetPackageVersionMetadata.cs:
```diff
@@ -0,0 +1,145 @@
+    /// <remarks>Necessary for unit test implementation of json serialization and deserialization.</remarks>
+    [JsonConstructor]
+#pragma warning disable CS8618
+    public NuGetPackageVersionMetadata()
```
So based on experimenting, and there could be another way of doing this, I believe I have to have this empty constructor in order to deserialize json. I can make it private and initialize all values as `null!` or set them to all be nullable to avoid the warning. But without this constructor, the unit tests at the json deserialization step fail.

gfs commented:
I finished the cargo manager implementation and refactored it into a TypedManager class that can be used by the actions and also refactored the Nuget manager.  Take a look at the TypedManager.
'
In this manner each project manager does need much additional code and the base functionality is tightly coupled to the provider interface.

What do you think?

gfs commented:
Here's an updated compare of whats changed: https://github.com/microsoft/OSSGadget/pull/323/files/9036d201ac002649d35e944f457930ebc90d3c87..3b2dd5e615044b4ec8e594ace89d599b2b086d72

gfs commented:
![image](https://user-images.githubusercontent.com/98900/161634257-107f1f24-ab90-4bba-a1f9-171b1c185f6b.png)

I made a little psudocode snippet for my proposal for the factory.

jpinz commented on src/Shared/Model/PackageActions/NuGetPackageActions.cs:
```diff
@@ -0,0 +1,155 @@
+    /// <inheritdoc />
+    public async Task<string> GetLatestVersionAsync(
+        PackageURL packageUrl,
+        bool includePrerelease = false,
```
Implemented. and added to GetAllVersionsAsync as well and implemented as well.

jpinz commented:
> I finished the cargo manager implementation and refactored it into a TypedManager class that can be used by the actions and also refactored the Nuget manager. Take a look at the TypedManager. ' In this manner each project manager does need much additional code and the base functionality is tightly coupled to the provider interface.
> 
> What do you think?

I love this. I'll go mess around with it myself, but makes total sense to me and is much cleaner.

gfs commented:
> > I finished the cargo manager implementation and refactored it into a TypedManager class that can be used by the actions and also refactored the Nuget manager. Take a look at the TypedManager. ' In this manner each project manager does need much additional code and the base functionality is tightly coupled to the provider interface.
> > What do you think?
> 
> I love this. I'll go mess around with it myself, but makes total sense to me and is much cleaner.

I also did most of the refactor for the factory as proposed above in another commit, heres the updated compare. It should be keyed on purl.type rather than purl.typeProjectManager. The mechanism by which you get the values to put for the initial default is indecided. It could be a static value in the managers that are being set? We could create a type that has a parameter that matches the purl type? Or just hardcode the strings in the default method.

https://github.com/microsoft/OSSGadget/pull/323/files/9036d201ac002649d35e944f457930ebc90d3c87..8d5a3c5b52b535226bb789f1c7bbac6ba119d05a

jpinz commented:
Going to make a third PR after #326 goes in that covers the changes between that one and this one.

jpinz commented:
Going to make a third PR after #326 goes in that covers the changes between that one and this one.
